### ISSUE TITLE
[tune] workers sit idle between training iterations when running on autoscaler

### ISSUE THREAD

**Issue by hartikainen** [ray-project/ray#4211]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04
- **Ray installed from (source or binary)**: Both
- **Ray version**: 0.6.2 and https://github.com/ray-project/ray/pull/4154/commits/2b623a118db889935a53cafe4f2441b8cf430fcc
- **Python version**: 3.6
- **Exact command to reproduce**: n/a

### Describe the problem
I'm seeing this issue where all my tune trials running on workers sit idle for most of the time. For example, in the following case, the cluster was started about 30 minutes before this log, and only the first two trials (that are running on the head node) are making reasonable progress. Rest of the trials are running about 1 step/10 minutes, even though based on the `time_total_s` it takes only 20 seconds per iteration.

```
RUNNING trials:
- id=f38ff9fe-seed=7678:       RUNNING [pid=3482], 1895 s, 58 iter, 58000 ts
- id=314e0483-seed=291:        RUNNING [pid=3483], 1897 s, 58 iter, 58000 ts
- id=3b333c47-seed=5763:       RUNNING [ray-20190301t094602-31c17c-dclaw3-screwv2-v0-worker-9aea164f pid=3161], 92 s, 4 iter, 4000 ts
- id=512ae016-seed=2978:       RUNNING [ray-20190301t094602-31c17c-dclaw3-screwv2-v0-worker-83ac45ed pid=3151], 93 s, 4 iter, 4000 ts
- id=8388ec55-seed=7037:       RUNNING [ray-20190301t094602-31c17c-dclaw3-screwv2-v0-worker-087e6b1d pid=3146], 92 s, 4 iter, 4000 ts
... 71 not shown
- id=5d69cff2-seed=1606:       RUNNING [ray-20190301t094602-31c17c-dclaw3-screwv2-v0-worker-40e01303 pid=3123], 79 s, 3 iter, 3000 ts
- id=0bf67558-seed=7161:       RUNNING [ray-20190301t094602-31c17c-dclaw3-screwv2-v0-worker-dcdda17b pid=3150], 77 s, 3 iter, 3000 ts
- id=2d8d3f48-seed=8881:       RUNNING [ray-20190301t094602-31c17c-dclaw3-screwv2-v0-worker-4774d905 pid=3192], 85 s, 3 iter, 3000 ts
- id=e805dfe9-seed=1388:       RUNNING [ray-20190301t094602-31c17c-dclaw3-screwv2-v0-worker-40e01303 pid=3119], 78 s, 3 iter, 3000 ts
- id=e1bc2581-seed=6844:       RUNNING [ray-20190301t094602-31c17c-dclaw3-screwv2-v0-worker-4774d905 pid=3193], 85 s, 3 iter, 3000 ts
```

Here are some diagnostics for one of the workers from the GCP console:
![image](https://user-images.githubusercontent.com/2308543/53661062-4ff54f80-3c14-11e9-8c5c-3055e3b3a9cd.png)

The cpu diagnostic shows that the cpus sit idle most of the, which matches well with the slowness showing up in the ray results. The same thing happens both on 0.6.2 and https://github.com/ray-project/ray/pull/4154/commits/2b623a118db889935a53cafe4f2441b8cf430fcc, and I don't see any problems when running things locally.

I'm not sure if this is a bug or maybe it's possible that for example the head node being slow could somehow stall the workers?

**Comment by ericl** [comment]
This sounds like https://github.com/ray-project/ray/issues/1128, but that test seems to pass now...

**Comment by hartikainen** [comment]
Some notes on this. It's kinda weird that I see this now both on 0.6.2 and https://github.com/ray-project/ray/commit/2b623a118db889935a53cafe4f2441b8cf430fcc even though I'm very confident that it has not happened ever before. The only thing that I can think of being different is that in the runs use median stopping rule, which I haven't used with autoscaler before.

Based on my discussion with @richardliaw, it seems likely that the cause for this is that when the master node does `ray.wait(...)` for a result, for some reason it's always the trials on the head node that get prioritized over worker nodes. At the same time, by design, the trials on the worker nodes stall while waiting their results to get recorded by the master.

I got around this issue temporarily by adding ray start option `--num-cpus=1` in the `head_start_commands` here https://github.com/ray-project/ray/blob/ray-0.6.2/python/ray/autoscaler/gcp/example-full.yaml#L153

**Comment by hartikainen** [comment]
Couple of follow up questions/thoughts related to my last comment:
- Is it necessary that the trials stall until their results get recorded to the master? I.e. would it be possible to buffer the results and keep the worker working and report multiple results at once?
- Is there something that should be done to the way the head node waits for the results? I.e. what stops us from gathering multiple results at the same time if they are ready?

**Comment by hartikainen** [comment]
Looking at the `ray.wait` docs, which says that:
> Ordering of the input list of object IDs is preserved. That is, if A precedes B in the input list, and both are in the ready list, then A will precede B in the ready list. This also holds true if A and B are both in the remaining list.

I wonder if shuffling the `ray.wait()` arguments in [RayTrialExecutor.get_next_available_trial](https://github.com/ray-project/ray/blob/bb7c4ce9c4807bb0344d11d9bd4fe656ae020014/python/ray/tune/ray_trial_executor.py#L219) would at least partially alleviate the issue?

**Comment by ericl** [comment]
One potential workaround to try is both shuffling the list and adding a small timeout to the wait. The timeout would make sure there isn't any case where the backend isn't returning trials on remote nodes if there existing local trials (this was the subject of the issue I linked).

**Comment by ericl** [comment]
Concretely:
```
--- a/python/ray/tune/ray_trial_executor.py
+++ b/python/ray/tune/ray_trial_executor.py
@@ -216,7 +216,9 @@ class RayTrialExecutor(TrialExecutor):
         return list(self._running.values())
 
     def get_next_available_trial(self):
-        [result_id], _ = ray.wait(list(self._running))
+        r = list(self._running)
+        random.shuffle(r)
+        [result_id], _ = ray.wait(r, num_returns=1, timeout=0.01)
         return self._running[result_id]
 
     def fetch_result(self, trial):
```

**Comment by ericl** [comment]
Thinking about it more, I'm still confused as to why shuffling makes such a big difference when the iteration time is O(tens of seconds).

Suppose a remote trial has finished a task. Even with full unfairness, that task should get returned as long as no local tasks are finished yet. It seems unlikely that a local task would be ready on every single wait (and in that case, the tune scheduler is the bottleneck and shuffling shouldn't make a difference).

**Comment by ericl** [comment]
an: I've also observed this same performance degradation around non-shuffled wait() on a single node only.

**Comment by stephanie-wang** [comment]
Hey @hartikainen, we merged #4242. Do you want to try it out? Thanks!

**Comment by max333** [comment]
I am seeing this problem too and it makes Tune unusable for me.  I don't think it has anything to do with shuffling the trial results requests.  I tested various versions and it started at 0.6.2.  The nightly build still has this bug, despite the shuffling fix.

The workers are idle most of the time and the cpu utilization is very low.

The problem is much more apparent when many jobs are scheduled.  I don't mean when many jobs are running in parallel, but just scheduled.  In the code I posted below, if you set `n_jobs = 64`, the problem is not perceptible.  But with `n_jobs = 2500`, only 1 of 8 running jobs is in state `IdleTraining.train()` from `ps` and the other seven are in the waiting state `IdleTraining`.

```
import time

from ray import tune
import ray.tune.util
from ray.tune import trainable


class IdleTrainable(trainable.Trainable):

    def _setup(self, config):
        self.name = config['name']
        self.timeout = config['timeout']

    def _train(self):
        time.sleep(self.timeout)
        string = f"hello from {self.name}"
        print(f"done: {self.name}")
        return {'done': True, 'message': string}

timeout = 3

def build_experiment(i):
    name = f"exp{i}"
    return tune.Experiment(
                run=IdleTrainable,
                name=name,
                config={"name": name,
                        "timeout": timeout},
                #num_samples=1, # We are using repeats instead.
                checkpoint_at_end=False,
                local_dir="/tmp/rayidle",
                # It seems this means a thread and not a full core.
                resources_per_trial={'cpu': 1})

ray.init()
n_jobs = 2500
experiments = [build_experiment(i) for i in range(n_jobs)]

tune.run_experiments(experiments, verbose=False)
```


I also ran the profiler:
```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   2352/1    0.074    0.000   66.300   66.300 {built-in method builtins.exec}
        1    0.000    0.000   66.299   66.299 junk_ray.py:1(<module>)
        1    0.000    0.000   62.646   62.646 tune.py:55(run_experiments)
       35    0.002    0.000   62.637    1.790 trial_runner.py:218(step)
       35    0.000    0.000   43.330    1.238 trial_runner.py:377(_get_next_trial)
       35    0.055    0.002   40.574    1.159 trial_scheduler.py:86(choose_trial_to_run)
    33375    0.028    0.000   40.519    0.001 trial_runner.py:373(has_resources)
    33375    0.089    0.000   40.491    0.001 ray_trial_executor.py:320(has_resources)
    33411    0.345    0.000   40.011    0.001 ray_trial_executor.py:292(_update_avail_resources)
   167051    0.558    0.000   39.222    0.000 state.py:750(cluster_resources)
   167051    0.232    0.000   38.638    0.000 state.py:390(client_table)
   167051    2.000    0.000   38.356    0.000 state.py:20(parse_client_table)
       34    0.004    0.000   17.156    0.505 trial_runner.py:127(checkpoint)
       68    3.035    0.045   17.107    0.252 __init__.py:120(dump)
```

It is spending the whole time in parse_client_table().  It is in state.py in the `experimental` module.  Hopefully that will ring a bell for someone.

**Comment by richardliaw** [comment]
For each Trainable object, can you time the amount of time in between `train` calls?

Also, are you using the autoscaler? If not, can you file a new issue? I think this issue was specifically tracking a multi-node bug.

**Comment by ericl** [comment]
I would be suspicious of any time lag in the TrialRunner::step method here: https://github.com/ray-project/ray/blob/1ce6f746bda802be03d54288ebaf7afa803cf0f2/python/ray/tune/trial_runner.py#L224

This PR adds some warn_if_slow blocks, but could probably cover other things like checkpoint(): https://github.com/ray-project/ray/pull/4218/files

**Comment by max333** [comment]
@ericl I showed a toy example here, but for our real workload each job lasts about 5 minutes.  The problem still appears when running those jobs over many nodes.  Maybe it's a different problem, but it looks very similar with nearly all jobs being in a waiting state.

**Comment by richardliaw** [comment]
I think we closed this issue with #4379 and #4302; feel free to reopen if otherwise.
